### PR TITLE
#40-6

### PR DESCRIPTION
**Pull request by 0leksandrr**
## Overview
This pull request introduces several security enhancements to the messaging system, focusing on content encryption, permission management, and rate limiting. These changes aim to protect user data and ensure the integrity of the messaging service.

## Key Changes

1. **Content Encryption**: 
   - Implemented encryption and decryption methods for message content, ensuring that sensitive information remains secure during storage and transmission.

2. **Input Sanitization and Validation**: 
   - Utilized Bandit to scan the project for potential security vulnerabilities, enhancing the overall security posture of the application.

3. **Permissions Management**: 
   - Established granular permissions to control user access to messages, ensuring that only authorized users can view or modify content.

4. **Rate Limiting**: 
   - Introduced rate limiting to prevent abuse and denial-of-service (DoS) attacks, controlling the number of messages that users can send within a specified timeframe.
 
Знімок екрана 2024-10-28 200121

### CODE REVIEW

**Review by uU-Uuu — APPROVED** [review]

File: communications/repositories/mongo.py [hunk 1]
```diff
@@ -14,25 +22,34 @@ class MongoDBRepository(BaseRepository):
...
+        ]
 
         logger.info(f"Creating chatroom with ID: {chatroom.room_id}")
 
         self._collection.insert_one(chatroom_dict)
 
         logger.info(f"Chatroom created successfully: {chatroom_dict}")
 
-    def get_chatroom(self, room_id: str) -> ChatRoom:
+    def get_chatroom(self, room_id: str) -> ChatRoom | None:
```
Comment (VovaD199) [inline communications/repositories/mongo.py:47]: def get_chatroom(self, room_id: str) -> Optional[ChatRoom]:

"| None" - looks like an unnecessary parameter

File: communications/repositories/mongo.py [hunk 2]
```diff
@@ -14,25 +22,34 @@ class MongoDBRepository(BaseRepository):
...
         logger.info(f"Retrieving chatroom with ID: {room_id}")
```
Comment (VovaD199) [inline communications/repositories/mongo.py:48]: try:

File: communications/repositories/mongo.py [hunk 3]
```diff
@@ -49,6 +66,8 @@ def add_message(self, room_id: str, message: Message):
         message_dict = message.__dict__
```
Comment (VovaD199) [inline communications/repositories/mongo.py:66]: add this code before the "def add_message" method:

        except PyMongoError as e:
            logger.error(f"Failed to retrieve chatroom: {e}", exc_info=True)
            return None

and everything in the "def add_message" method must be wrapped in " try.... except"

for example:
except PyMongoError as e:
            logger.error(f"Failed to add message to MongoDB: {e}", exc_info=True)
        except Exception as e:
            logger.error(f"Unexpected error during message encryption or database update: {e}", exc_info=True)

File: communications/views.py [hunk 4]
```diff
@@ -14,56 +21,67 @@ class CreateChatRoomView(APIView):
...
+
         request_data = request.data.copy()
         request_data['room_id'] = room_id
         if 'sender_id' not in request_data:
+            logger.error("sender_id is required in the request data")
             return Response({'error': 'sender_id is required.'}, status=status.HTTP_400_BAD_REQUEST)
 
         serializer = MessageSerializer(data=request_data)
         if serializer.is_valid():
             message = serializer.create(serializer.validated_data)
```
Comment (VovaD199) [inline communications/views.py:57]: try:

File: communications/views.py [hunk 5]
```diff
@@ -14,56 +21,67 @@ class CreateChatRoomView(APIView):
...
             mongo_container.add_message(room_id, message)
+            logger.info(f"Message sent with ID: {message.oid} in room: {room_id}")
             return Response({'message_id': str(message.oid)}, status=status.HTTP_201_CREATED)
```
Comment (VovaD199) [inline communications/views.py:61]: except Exception as e:
                logger.error(f"Failed to send message in room {room_id}: {e}", exc_info=True)
                return Response(
                    {'error': 'Failed to send message due to server error.'},
                    status=status.HTTP_500_INTERNAL_SERVER_ERROR
                )

File: communications/views.py [hunk 6]
```diff
@@ -14,56 +21,67 @@ class CreateChatRoomView(APIView):
...
     """
     View for listing all messages in a specific chat room.
     """
+    permission_classes = (IsOwnerOrRecipient,)
 
+    @ratelimit(key='ip', rate='10/m', method='GET', block=True)
     def get(self, request, room_id):
-        """
-        Retrieve all messages in a specific chat room, sorted by creation date.
+        logger.info(f"ListMessagesView GET request received for room_id: {room_id}")
```
Comment (VovaD199) [inline communications/views.py:75]: try:

File: communications/views.py [hunk 7]
```diff
@@ -14,56 +21,67 @@ class CreateChatRoomView(APIView):
...
             message_list = sorted(
                 (asdict(msg) for msg in chat_room.messages),
                 key=lambda msg: msg.get('created_at'),
                 reverse=False
             )
+            logger.info(f"Retrieved {len(message_list)} messages for room_id: {room_id}")
             return Response(message_list, status=status.HTTP_200_OK)
 
+        logger.error(f"Chat room not found for room_id: {room_id}")
         return Response({'error': 'Chat room not found.'}, status=status.HTTP_404_NOT_FOUND)
```
Comment (VovaD199) [inline communications/views.py:87]: except Exception as e:
            logger.error(f"Failed to retrieve chat room {room_id}: {e}", exc_info=True)
            return Response(
                {'error': 'Failed to retrieve messages due to server error.'},
                status=status.HTTP_500_INTERNAL_SERVER_ERROR
            )

**Review by VovaD199 — APPROVED** [review]

File: communications/serializers.py [hunk 8]
```diff
@@ -13,6 +14,18 @@ class MessageSerializer(serializers.Serializer):
     content = serializers.CharField(max_length=1000)
     read_at = serializers.DateTimeField(required=False, allow_null=True)
 
+    def validate_content(self, value):
+        """Custom validation for content field to check for XSS vulnerabilities."""
+        if "<script>" in value or "javascript:" in value:
+            raise serializers.ValidationError("Content contains potentially dangerous XSS code.")
+
+        sanitized_content = bleach.clean(value, tags=[], attributes={}, styles=[], strip=True)
+
```
[the remark this answers is not in the document]
Comment (0leksandrr) [reply communications/serializers.py]: Done

File: communications/repositories/mongo.py [hunk 9]
```diff
@@ -1,11 +1,19 @@
+import logging
 from dataclasses import dataclass
+from typing import Any, Mapping
```
[the remark this answers is not in the document]
Comment (0leksandrr) [reply communications/repositories/mongo.py]: Done

File: communications/repositories/mongo.py [hunk 10]
```diff
@@ -1,11 +1,19 @@
...
+
+from cryptography.fernet import Fernet
+from django.conf import settings
 from pymongo import MongoClient
-import logging
-from .base import BaseRepository
+from pymongo.synchronous.collection import Collection
+
```
[the remark this answers is not in the document]
Comment (0leksandrr) [reply communications/repositories/mongo.py:10]: Done

File: communications/repositories/mongo.py [hunk 11]
```diff
@@ -14,25 +22,34 @@ class MongoDBRepository(BaseRepository):
     mongo_db_collection_name: str
 
     @property
-    def _collection(self):
+    def _collection(self) -> Collection[Mapping[str, Any] | Any]:
```
[the remark this answers is not in the document]
Comment (0leksandrr) [reply communications/repositories/mongo.py]: Done

File: communications/repositories/mongo.py (hunk 3 above)
Comment (0leksandrr) [reply communications/repositories/mongo.py:77]: Done

File: communications/views.py (hunk 4 above)
Comment (0leksandrr) [reply communications/views.py]: Done

File: communications/views.py (hunk 5 above)
Comment (0leksandrr) [reply communications/views.py]: Done

File: communications/views.py (hunk 6 above)
Comment (0leksandrr) [reply communications/views.py]: Done

File: communications/serializers.py [hunk 12]
```diff
@@ -13,6 +14,18 @@ class MessageSerializer(serializers.Serializer):
...
+        if not sanitized_content:
+            raise serializers.ValidationError("Content cannot be empty after sanitization.")
+
+        return sanitized_content
+
     def create(self, validated_data):
         """Create a new Message instance from validated data and save it in MongoDB."""
         message = Message(**validated_data)
```
[the remark this answers is not in the document]
Comment (0leksandrr) [reply communications/serializers.py:34]: good advice about using context. Done

File: communications/views.py (hunk 7 above)
Comment (0leksandrr) [reply communications/views.py]: Done

**Review by SaberanPWNZ — APPROVED** [review]

**Pull request merged**